src/Symfony/Component/HttpClient/Rfc9111HttpClient.php

```diff
@@ -0,0 +1,616 @@
+
+                            return;
+                        }
+                    }
+
+                    // recomputing vary fields in case it changed or for first request
+                    $varyFields = [];
+                    foreach ($headers['vary'] ?? [] as $vary) {
+                        foreach (explode(',', $vary) as $field) {
+                            $varyFields[] = trim($field);
+                        }
+                    }
+
+                    $metadataKey = $this->getMetadataKey($requestHash, $options['normalized_headers'], $varyFields);
+
+                    yield $chunk;
+
+                    return;
+                }
+
+                if (!$this->isServerResponseCacheable($context->getStatusCode(), $options['normalized_headers'], $headers, $cacheControl)) {
+                    $context->passthru();
+
+                    yield $chunk;
+
+                    return;
+                }
+
+                if ($chunk->isLast()) {
+                    $this->cache->saveDeferred($varyItem->set($varyFields)->tag($fullUrlTag)->expiresAt($expiresAt));
+
+                    $maxAge = $this->determineMaxAge($headers, $cacheControl);
+
+                    $this->cache->saveDeferred(
+                        $this->cache->getItem($metadataKey)
+                            ->tag($fullUrlTag)
+                            ->set([
+                                'status_code' => $context->getStatusCode(),
+                                'headers' => $headers,
+                                'initial_age' => (int) ($headers['age'][0] ?? 0),
+                                'stored_at' => time(),
+                                'expires_at' => $this->calculateExpiresAt($maxAge),
+                                'chunks_count' => $chunkIndex,
+                            ])
+                            ->expiresAt($expiresAt)
+                    );
+
+                    $this->cache->commit();
+
+                    yield $chunk;
+
+                    return;
+                }
+
+                ++$chunkIndex;
+                $chunkKey = "{$metadataKey}_chunk_{$chunkIndex}";
+                $chunkItem = $this->cache->getItem($chunkKey)
+                    ->tag($fullUrlTag)
+                    ->set($chunk->getContent())
+                    ->expiresAfter($this->ttl);
+
+                $this->cache->save($chunkItem);
+
+                yield $chunk;
+            }
+        );
+    }
+
+    public function stream(ResponseInterface|iterable $responses, ?float $timeout = null): ResponseStreamInterface
+    {
+        if ($responses instanceof ResponseInterface) {
+            $responses = [$responses];
+        }
+
+        $mockResponses = [];
+        $asyncResponses = [];
+        $clientResponses = [];
+
+        foreach ($responses as $response) {
+            if ($response instanceof MockResponse) {
+                $mockResponses[] = $response;
+            } elseif ($response instanceof AsyncResponse) {
```

Comment:
What happens if the inner client also uses an AsyncResponse, and we are in the case where the `request` method returns the inner response directly ?

Conditional return of a forwarded response or an owned response makes it hard to implement streaming reliably.